src/Kestrel.Transport.Sockets/Internal/IOQueue.cs

```diff
@@ -24,35 +28,69 @@ public override void Schedule(Action<object> action, object state)
                 State = state
             };
 
+            // Order is important here with Execute.
+            // Enqueue prior to checking _doingWork.
             _workItems.Enqueue(work);
 
-            lock (_workSync)
+            // Ensure ordering of write -> read is preserved, as order is reversed between Schedule and Execute,
+            // and they are two different memory locations.
+            Thread.MemoryBarrier(); 
+
+            if (Volatile.Read(ref _doingWork) == 0)
             {
-                if (!_doingWork)
+                // Set as working, and check if it was already working.
+                var submitWork = Interlocked.Exchange(ref _doingWork, 1) == 0;
+
+                if (submitWork)
                 {
+                    // Wasn't scheduled or active, schedule.
+#if NETCOREAPP3_0
+                    System.Threading.ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+#else
                     System.Threading.ThreadPool.UnsafeQueueUserWorkItem(_doWorkCallback, this);
-                    _doingWork = true;
+#endif
                 }
             }
         }
 
-        private void DoWork()
+#if NETCOREAPP3_0
+        void IThreadPoolWorkItem.Execute()
+#else
+        private void Execute()
+#endif
         {
             while (true)
             {
-                while (_workItems.TryDequeue(out Work item))
+                var workItems = _workItems;
+                while (workItems.TryDequeue(out Work item))
                 {
                     item.Callback(item.State);
                 }
 
-                lock (_workSync)
+                // Order is important here with Schedule.
+                // Set _doingWork prior to checking .IsEmpty
+                Volatile.Write(ref _doingWork, 0);
+
+                // Ensure ordering of write -> read is preserved, as order is reversed between Schedule and Execute,
+                // and they are two different memory locations.
+                Thread.MemoryBarrier();
```

Comment:
@stephentoub seem to have a race condition that only happens on netfx without the `Thread.MemoryBarrier` (added x2 to be sure); could this be due to the different implementations of `ConcurrentQueue<T>` on .NET Core and Full Framework and `workItems.IsEmpty` below only acting as a barrier on .NET Core?

Reply:
> and workItems.IsEmpty below only acting as a barrier on .NET Core?

I doubt it.  Both implementations perform volatile reads, and if memory serves neither implementation has any volatile writes or full barriers.

Reply:
Probably just that am depending on the ordering of read and writes to two different memory locations then

Reply:
Probably just that am depending on the ordering of read and writes to two different memory locations then. Will drop `Volatile` since using `MemoryBarrier`